### PR TITLE
Add support for configuring the build system scheduler

### PR DESCRIPTION
**Pull request by dmbryson**
- Add methods for changing the scheduler algorithm. Supports:
-- commandNamePriority: priority queue ordered by command name [default]
-- fifo: first in/first out queue - previous behavior
- Add methods for setting the lane width of the scheduler
- Add command line option for llbuild debug tool for selecting the scheduler

### CODE REVIEW

**Comment by dmbryson** [comment]
@swift-ci please smoke test

File: include/llbuild/BuildSystem/BuildSystem.h [hunk 1]
```diff
@@ -295,6 +295,22 @@ enum class QualityOfService {
...
 
+// MARK: Execution Queue Scheduler Control
+
+enum class SchedulerAlgorithm {
+  /// Command name priority queue based scheduling [default]
+  commandNamePriority = 0,
+
+  /// First in, first out
+  fifo = 1
+};
```
Comment (ddunbar) [inline include/llbuild/BuildSystem/BuildSystem.h]: Unlike Swift, we have been using initial-uppercase for `enum class` constants... 🤷‍♂️

File: products/libllbuild/BuildSystem-C-API.cpp [hunk 2]
```diff
@@ -811,7 +811,7 @@ char* llb_buildsystem_command_get_verbose_description(
   return strdup(result.c_str());
 }
 
-llb_quality_of_service_t llb_get_quality_of_servicellb_enable_tracing() {
+llb_quality_of_service_t llb_get_quality_of_service() {
```
Comment (ddunbar) [inline products/libllbuild/BuildSystem-C-API.cpp]: Thanks! 😊

File: lib/BuildSystem/LaneBasedExecutionQueue.cpp [hunk 3]
```diff
@@ -607,3 +662,18 @@ void llbuild::buildsystem::setDefaultQualityOfService(QualityOfService level) {
...
+void llbuild::buildsystem::setSchedulerAlgorithm(SchedulerAlgorithm algorithm) {
+  defaultAlgorithm = algorithm;
+}
+
+uint32_t llbuild::buildsystem::getSchedulerLaneWidth() {
+  return defaultLaneWidth;
+}
+
+void llbuild::buildsystem::setSchedulerLaneWidth(uint32_t width) {
+  defaultLaneWidth = width;
```
Comment (ddunbar) [inline lib/BuildSystem/LaneBasedExecutionQueue.cpp]: I wonder if we should impose a limit here? Unlike systems like say Ninja which uses a select-based mechanism, our lane based approach isn't very compatible with setting a super high lane count... OTOH, its hard to pick a sane limit, so maybe we should just say caveat emptor.

File: include/llbuild/BuildSystem/BuildSystem.h (hunk 1 above)
Comment (dmbryson) [reply include/llbuild/BuildSystem/BuildSystem.h]: Ah, I matched the style of QualityOfService right above.  😉

Probably worth cleaning up at some point, but I'll save it for later <rdar://problem/39958068>

File: lib/BuildSystem/LaneBasedExecutionQueue.cpp (hunk 3 above)
Comment (dmbryson) [reply lib/BuildSystem/LaneBasedExecutionQueue.cpp]: Caveat emptor it is

File: include/llbuild/BuildSystem/BuildSystem.h (hunk 1 above)
Comment (ddunbar) [reply include/llbuild/BuildSystem/BuildSystem.h]: Well played

**Pull request merged**